### PR TITLE
Enable text-generation with new API

### PR DESCRIPTION
**Pull request by changwangss**
## Type of Change

```
    from intel_extension_for_transformers.transformers import (
        MixedPrecisionConfig,
        WeightOnlyQuantConfig,
        SmoothQuantConfig,
        BitsAndBytesConfig

    ) 
    from intel_extension_for_transformers.transformers import AutoModelForCausalLM
    # smooth-quant
    sq_config = SmoothQuantConfig(
                                tokenizer=tokenizer,  # either two of one, tokenizer or calib_func
                               )
    q_model = AutoModelForCausalLM.from_pretrained(model_name_or_path,
                                                   quantization_config=sq_config
                                               )
    
    # weight-only
    woq_config = WeightOnlyQuantConfig()
    woq_model = AutoModelForCausalLM.from_pretrained(model_name_or_path,
                                                quantization_config=woq_config
                                            )
    
    # mp
    mp_config = MixedPrecisionConfig() 
    amp_model = AutoModelForCausalLM.from_pretrained(model_name_or_path,
                                                quantization_config=mp_config
                                            )
  
    # bitsandbytes
    bab_config = BitsAndBytesConfig()
    bab_model = AutoModelForCausalLM.from_pretrained(model_name_or_path,
                                                quantization_config=bab_config
                                            )
```

## Description

detail description 
JIRA ticket: https://jira.devtools.intel.com/browse/NLPTOOLKIU-878

## Expected Behavior & Potential Risk

the expected behavior that triggered by this PR 

## How has this PR been tested?

how to reproduce the test (including hardware information)

## Dependency Change?

any library dependency introduced or removed

### CODE REVIEW

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py [hunk 1]
```diff
@@ -17,6 +17,9 @@
 import numpy as np
 from itertools import chain
 from optimum.utils import NormalizedConfigManager
+# ipex dependency
```
Comment (hshen14) [inline examples/huggingface/pytorch/text-generation/quantization/run_generation.py:20]: do you need this change? assuming these would be hided into llm API?

File: intel_extension_for_transformers/neural_chat/config.py [hunk 2]
```diff
@@ -417,6 +417,14 @@ class WeightOnlyQuantizationConfig:
 @dataclass
 class AMPConfig:
     dtype: str = 'bfloat16'
+    op_type_dict = None
```
Comment (hshen14) [inline intel_extension_for_transformers/neural_chat/config.py:420]: why need this for AMP?

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py (hunk 1 above)
Comment (changwangss) [reply examples/huggingface/pytorch/text-generation/quantization/run_generation.py:20]: The SmoothQuantConfig backend is ipex only, so it is necessary import IPEX, yes, I can hide it to llm API.

File: intel_extension_for_transformers/neural_chat/config.py (hunk 2 above)
Comment (changwangss) [reply intel_extension_for_transformers/neural_chat/config.py:420]: wrong typing. will remove.

**Review by hshen14 — APPROVED** [review]

File: intel_extension_for_transformers/llm/quantization/optimization.py [hunk 3]
```diff
@@ -55,4 +66,139 @@ def optimize(self, model):
...
+                example_inputs=example_inputs,
+            )
+            if calib_func is None:
+                logger.info("The default calibration funcation is used, " +
+                            "the calibration dataset is NeelNanda/pile-10k," +
+                            "batchsize is 1 and calibration iteration is 100.")
+                calib_func = default_calib_func
+            else:
+                calib_func = calib_func
+            optimized_model = quantization.fit(
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/quantization/optimization.py:148]: INC API?

**Review by kevinintel — APPROVED** [review]

**Review by XinyuYe-Intel — APPROVED** [review]

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py [hunk 4]
```diff
@@ -58,198 +55,93 @@
...
-        )
-    # save config
-    user_model.config.save_pretrained(args.output_dir)
-    q_model = quantization.fit(
-        user_model,
-        conf,
-        calib_func=calib_func,
-    )
-    q_model.save(args.output_dir)
+    sq_config = SmoothQuantConfig(
```
[the remark this answers is not in the document]
Comment (changwangss) [reply examples/huggingface/pytorch/text-generation/quantization/run_generation.py:110]: it is the temporary format for validation, if the API  is ready, I will update the example and add ut.

File: intel_extension_for_transformers/transformers/utils/quantization_config.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    enable_full_range: bool = True
+
+@dataclass
+class AMPConfig:
+    dtype: str = 'bfloat16'    
+
+@dataclass
+class SmoothQuantConfig:
+    tokenizer: Any = None
+    calib_func: Any = None
```
Comment (ftian1) [inline intel_extension_for_transformers/transformers/utils/quantization_config.py:39]: 1. why there is no dataset related setting exposed and hard code in internal? 
2. calib_func is quite ugly. is possible to remove it? FX tracing/ multiple modules issue shouldn't introduce such argument and bring complexity.

**Review by hshen14 — APPROVED** [review]

File: intel_extension_for_transformers/transformers/utils/quantization_config.py (hunk 5 above)
Comment (changwangss) [reply intel_extension_for_transformers/transformers/utils/quantization_config.py:39]: reply Q1:
for built-in calib_func, I don't know what key for each sample if we open the args.dataset. eg: example["text"], example["code"]
reply Q2:
I think the default calib_func is calib_func is mainly aimed at LLM quantization, for other models such as qa task specifically is not suitable, user need provided calib_func by themselves.

**Comment by changwangss** [comment]
confilct with https://github.com/intel/intel-extension-for-transformers/pull/297, need align with @PenghuiCheng

**Comment by hshen14** [comment]
> confilct with #297, need align with @PenghuiCheng

what's the conflict?

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py [hunk 6]
```diff
@@ -58,198 +55,93 @@
...
-                past_key_values=past_key_values,
-                attention_mask=attention_mask,
-            )
-
-    from neural_compressor import PostTrainingQuantConfig, quantization
-
-    if re.search("gptj", user_model.config.model_type) or re.search(
-        "gpt_neox", user_model.config.model_type
+    from intel_extension_for_transformers.transformers import (
+        AMPConfig,
```
Comment (hshen14) [inline examples/huggingface/pytorch/text-generation/quantization/run_generation.py:89]: use MixedPrecisionConfig

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py [hunk 7]
```diff
@@ -58,198 +55,93 @@
...
+        WeightOnlyQuantizationConfig,
```
Comment (hshen14) [inline examples/huggingface/pytorch/text-generation/quantization/run_generation.py:90]: Use WeightOnlyQuantConfig

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py [hunk 8]
```diff
@@ -58,198 +55,93 @@
...
+    amp_model = AutoModelForCausalLM.from_pretrained(args.model,
+                                                quantization_config=amp_config
+                                            )
+    print("amp done.")
+    # bitsandbytes
+    bab_config = BitsAndBytesConfig()
+    bab_model = AutoModelForCausalLM.from_pretrained(args.model,
+                                                quantization_config=bab_config
+                                            )
+    print("bitsandbytes done.")
```
Comment (hshen14) [inline examples/huggingface/pytorch/text-generation/quantization/run_generation.py:143]: Why we need to use these configs/generate these models one by one? they should be enabled based on parameter?

File: intel_extension_for_transformers/transformers/utils/utility.py [hunk 9]
```diff
@@ -30,6 +30,7 @@
 DECODER_WITH_PAST_NAME = "decoder_with_past_model.bin"
 WEIGHTS_NAME = "pytorch_model.bin"
 
+torch = LazyImport("torch")
```
Comment (hshen14) [inline intel_extension_for_transformers/transformers/utils/utility.py:33]: better to rename the file to "generation_utility.py"?

File: intel_extension_for_transformers/transformers/utils/quantization_config.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+"""Configs for intel extension for transformers."""
+
+from dataclasses import dataclass, field
+from typing import Any, Optional
+
+from transformers import BitsAndBytesConfig
+
+
+@dataclass
+class WeightOnlyQuantizationConfig:
```
Comment (hshen14) [inline intel_extension_for_transformers/transformers/utils/quantization_config.py:26]: WeightOnlyQuantConfig

File: intel_extension_for_transformers/transformers/utils/quantization_config.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    algorithm: str = "RTN"
+    bits: int = 8
+    group_size: int = -1
+    scheme: str = "sym"
+    enable_full_range: bool = True
+
+
+@dataclass
+class AMPConfig:
```
Comment (hshen14) [inline intel_extension_for_transformers/transformers/utils/quantization_config.py:35]: MixedPrecisionConfig

File: examples/huggingface/pytorch/text-generation/quantization/run_generation.py (hunk 8 above)
Comment (changwangss) [reply examples/huggingface/pytorch/text-generation/quantization/run_generation.py:143]: it seems not only have generate utility but also have dataloader and distribute utility. @hshen14 @PenghuiCheng

File: intel_extension_for_transformers/transformers/utils/utility.py (hunk 9 above)
Comment (changwangss) [reply intel_extension_for_transformers/transformers/utils/utility.py:33]: it seems not only have generate utility but also have dataloader and distribute utility. @hshen14 @PenghuiCheng

File: examples/huggingface/pytorch/text-generation/quantization/README.md [hunk 12]
```diff
@@ -6,106 +6,101 @@ We provide the inference benchmarking script `run_generation.py` for [EleutherAI
...
-    --int8_bf16_mixed \
-    --ipex
+    --benchmark
+# mixedprecision
+OMP_NUM_THREADS=<physical cores num> numactl -m <node N> -C <cpu list> python run_generation.py \
+    --model EleutherAI/gpt-j-6b \
+    --mixed_precision \
+    --benchmark
+# smoothquant
+# --int8_bf16_mixed is used for int8 mixed bfloat16 precision.
```
Comment (kevinintel) [inline examples/huggingface/pytorch/text-generation/quantization/README.md:52]: The parameter is not in example, remove it

File: examples/huggingface/pytorch/text-generation/quantization/README.md [hunk 13]
```diff
@@ -6,106 +6,101 @@ We provide the inference benchmarking script `run_generation.py` for [EleutherAI
...
+    --accuracy \
+    --tasks "lambada_openai"
+# mixedprecision
+python run_generation.py \
+    --model EleutherAI/gpt-j-6b \
+    --mixed_precision \
+    --accuracy \
+    --tasks "lambada_openai"
+# smoothquant
 # --int8_bf16_mixed is used for int8 mixed bfloat16 precision.
```
Comment (kevinintel) [inline examples/huggingface/pytorch/text-generation/quantization/README.md:86]: you already removed this parameter

**Pull request merged**